File: scripts/fetch_all_data.py
import json
import math
import os
import time

from openai import OpenAI

# Настройка OpenRouter
client = OpenAI(api_key="YOUR_OPENROUTER_KEY")

cities = [
    {"name": "Tokyo", "country": "Japan"},
    {"name": "Delhi", "country": "India"},
    {"name": "Shanghai", "country": "China"},
    {"name": "São Paulo", "country": "Brazil"},
    {"name": "Mexico City", "country": "Mexico"},
    {"name": "Cairo", "country": "Egypt"},
    {"name": "Mumbai", "country": "India"},
    {"name": "Beijing", "country": "China"},
    {"name": "Dhaka", "country": "Bangladesh"},
    {"name": "Osaka", "country": "Japan"},
    {"name": "New York", "country": "USA"},
    {"name": "Karachi", "country": "Pakistan"},
    {"name": "Buenos Aires", "country": "Argentina"},
    {"name": "Chongqing", "country": "China"},
    {"name": "Istanbul", "country": "Turkey"},
    {"name": "Kolkata", "country": "India"},
    {"name": "Manila", "country": "Philippines"},
    {"name": "Lagos", "country": "Nigeria"},
    {"name": "Rio de Janeiro", "country": "Brazil"},
    {"name": "Tianjin", "country": "China"}
]

metrics = [
    "homicide",
    "traffic_deaths",
    "incarceration",
    "suicide",
    "infant_mortality",
    "hiv",
    "freedom",
    "economic_freedom",
    "corruption",
    "hdi",
    "pisa",
    "english_speakers",
    "languages"
]

ranges = {
    "homicide": (0, 100),
    "traffic_deaths": (0, 100),
    "incarceration": (0, 2000),
    "suicide": (0, 50),
    "infant_mortality": (0, 200),
    "hiv": (0, 50000000),
    "freedom": (0, 100),
    "economic_freedom": (1900, 2030),
    "corruption": (0, 100),
    "hdi": (0, 1),
    "pisa": (0, 600),
    "english_speakers": (0, 2000000000),
    "languages": (0, 1000)
}

CACHE_FILE = "city_data_cache.json"

def parse_numeric(value, key=None):
    """Преобразуем текстовое значение в число и проверяем диапазон"""
    try:
        num = float(value)
        if math.isnan(num) or math.isinf(num):
            return None
        if key and key in ranges:
            low, high = ranges[key]
            if not (low <= num <= high):
                return None
        return num
    except:
        return None

def fetch_city_data(city_name, country_name):
    """Запрос к LLM для получения данных по городу и стране"""
    prompt = f"""
    Provide the following statistics for the city "{city_name}" in "{country_name}":
    - homicide rate per 100,000
    - traffic deaths per 100,000
    - incarceration rate per 100,000
    - suicide rate per 100,000
    - infant mortality per 1,000
    - HIV cases (total number)
    - freedom index (0-100)
    - economic freedom score (year)
    - corruption perception index (0-100)
    - HDI (0-1)
    - PISA score (0-600)
    - English speakers (number in city)
    - Languages spoken (number)

    Return only JSON with keys exactly as above, using null for unknown or out-of-range values.
    """

    for attempt in range(3):
        try:
            response = client.chat.completions.create(
                model="gpt-5-mini",
                messages=[{"role": "user", "content": prompt}],
                temperature=0
            )
            text = response.choices[0].message.content
            data = json.loads(text)
            for key in data:
                if key in metrics:
                    data[key] = parse_numeric(data[key], key)
            return data
        except Exception as e:
            print(f"⚠️ Ошибка при получении данных для {city_name}: {e}")
            time.sleep(2)
    return {k: None for k in metrics}

def load_cache():
    if os.path.exists(CACHE_FILE):
        with open(CACHE_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    return {}

def save_cache(cache):
    with open(CACHE_FILE, "w", encoding="utf-8") as f:
        json.dump(cache, f, indent=2, ensure_ascii=False)

def main():
    cache = load_cache()
    all_data = {}
    for city in cities:
        name = city["name"]
        if name in cache:
            print(f"ℹ️ Используем кэшированные данные для {name}")
            data = cache[name]
        else:
            print(f"🌍 Загружаем данные для {name} ({city['country']}) ...")
            data = fetch_city_data(name, city["country"])
            cache[name] = data
            save_cache(cache)
            time.sleep(1)  # пауза между запросами
        all_data[name] = data

    with open("city_data.json", "w", encoding="utf-8") as f:
        json.dump(all_data, f, indent=2, ensure_ascii=False)

    print("✅ Все данные загружены и сохранены в city_data.json")

if __name__ == "__main__":
    main()

File: scripts/fetch_all_data.js
const fs = require("fs");
const path = require("path");
const axios = require("axios");
require("dotenv").config();

const apiKey = process.env.OPENROUTER_API_KEY;
if (!apiKey) {
  console.error("❌ Нет API ключа! Добавь его в .env");
  process.exit(1);
}

const jsonFilePath = path.join(__dirname, "../public/data/cities_final_with_continent.json");

let cityData = [];
try {
  if (fs.existsSync(jsonFilePath)) {
    const raw = fs.readFileSync(jsonFilePath, "utf-8");
    cityData = JSON.parse(raw);
  }
} catch (err) {
  console.error("Ошибка при чтении JSON:", err.message);
}

// Функция запроса к LLM
async function fetchCityData(cityName) {
  const prompt = `
Provide strictly valid JSON (no extra text) with the following fields for the city of ${cityName}:
cost_of_living_index (number),
average_salary (number),
safety_index (number),
healthcare_index (number),
pollution_index (number),
climate_index (number),
quality_of_life_index (number),
education_index (number),
transport_index (number),
internet_speed (number),
english_proficiency (number),
population (number),
life_expectancy (number)
Return only JSON, nothing else.
`;

  try {
    const response = await axios.post(
      "https://openrouter.ai/api/v1/chat/completions",
      {
        model: "cognitivecomputations/dolphin3.0-r1-mistral-24b:free",
        messages: [{ role: "user", content: prompt }],
        temperature: 0,
      },
      {
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
      }
    );

    let content = response.data.choices[0].message.content;

    // Очистка кавычек
    content = content.replace(/[\u201C\u201D]/g, '"').replace(/[\u2018\u2019]/g, "'");

    try {
      return JSON.parse(content);
    } catch (parseErr) {
      console.error(`Ошибка парсинга JSON для ${cityName}:`, parseErr.message);
      console.log("Ответ LLM:", content);
      return null;
    }
  } catch (err) {
    console.error(`Ошибка при получении данных для ${cityName}:`, err.message);
    return null;
  }
}

async function main() {
  if (!Array.isArray(cityData) || cityData.length === 0) {
    console.log("⚠️ JSON пустой или не массив. Добавьте хотя бы один объект города.");
    return;
  }

  for (const cityObj of cityData) {
    const cityName = cityObj.name;
    console.log(`⏳ Обрабатываю ${cityName}...`);
    const llmData = await fetchCityData(cityName);
    if (llmData) {
      Object.assign(cityObj, llmData); // Добавляем новые поля к существующему объекту
      console.log(`✅ Данные для ${cityName} обновлены`);
    }
    await new Promise((r) => setTimeout(r, 2000)); // Пауза 2 сек
  }

  fs.writeFileSync(jsonFilePath, JSON.stringify(cityData, null, 2), "utf-8");
  console.log(`\n🎉 Все данные сохранены в ${jsonFilePath}`);
}

main();
